### PR TITLE
Expose functions from `AudioStreamPlaybackMicrophone` to access the microphone `AudioDriver::input_buffer` independently of the rest of the Audio system

### LINKED ISSUE #11347: A new InputEventMicrophone class to deliver audio chunks of microphone data independently of the Audio system
### Describe the project you are working on

A reliable and easy to use Voice over IP plugin https://github.com/goatchurchprime/two-voip-godot-4

### Describe the problem or limitation you are having in your project

The microphone input is unreliable and sounds distorted on different platforms when it is put to use.  

I have gone into the implementation in depth as discussed in my [previous proposal](https://github.com/godotengine/godot-proposals/issues/11325#issuecomment-2539505337).  

TDLR: 
* When the rate of the input samples from microphone does not exactly match the rate of the output sample in the audio system the buffering between them fails and causes bugs.  
* The current implementation uses an extra buffer that causes lag in the audio encoding.  
* Microphone data is not required in the audio system, as evidenced by the fact that it must always be routed to a bus that is set to mute.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The microphone would be input in the same way as the midi and the mouse is input, and can be handled in the same way.

The most common purpose of microphone data is for packaging and compressing into Opus packets for transmission across the network to another player.  The unpack and playback stage in the other player's game has to handle the jitter and buffering of packets as well the mismatch between the clockrates on different computers.  Compared to this, any jittering on the rate of input is irrelevant. 

The second most common purpose of microphone data is as a trigger (shout or blowing in the mic) to set off an event.

The least common use is to feed the sound into the speakers because it causes a feedback screech. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Basing this on the [Midi Input](https://docs.godotengine.org/en/stable/classes/class_inputeventmidi.html#class-inputeventmidi) we should control this with the functions:

* [AudioServer.get_input_device_list()](https://docs.godotengine.org/en/stable/classes/class_audioserver.html#class-audioserver-method-get-input-device-list)
* AudioServer.start_microphone_input(chunk_size: int, input_device: string)
* AudioServer.stop_microphone_input()

Then we emit an `InputEventMicrophone` every time the [AudioDriver::input_buffer](https://github.com/godotengine/godot/blob/master/servers/audio_server.h#L62) has more than `chunk_size` samples in it, with the properties:
 
* data : PackedVector2Array   # length = chunk_size
* abs_max : float   # maximum volume of the samples in the chunks

These packets could be resampled and encoded into opus chunks the moment they are emitted to the `_input()` function, and not in a process loop that has to poll the buffer to check if it's got enough samples to form a chunk.  The default Opus chunk size is 20ms long which is 882 samples at 44.1kHz.  

This interface also very useful if you want to use a voice activated sound (eg shout or blow into the microphone to fire a bullet) by monitoring the value of abs_max as you would the events from a pressure sensitive pad.  

To avoid copying large chunks of data around in the InputEventMicrophone, we could instead return indexes into the `input_buffer` and copy the data to the Opus encoding buffer with a function:

* AudioServer.copy_input_data(PackedVector2Array&, start, length)

The code required for this implementation of this feature would be modest.  It is the case that [all five platforms](https://github.com/godotengine/godot-proposals/issues/11325#issuecomment-2539505337) fill the input_buffer via the [AudioDriver::input_buffer_write(int32_t sample)](https://github.com/godotengine/godot/blob/master/servers/audio_server.cpp#L105) function in chunks of varying size.  At the end of each of these chunks we need to call a new function:
* AudioDriver::generate_microphone_events()

This would construct as many events in the buffer as necessary that matched the chunk_size, and then call `Input::get_singleton()->parse_input_event(event)` on each one to enter it into the system.
  

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

The microphone is core.

### PR DESCRIPTION
**Pull request by goatchurchprime**
This solves all the issues raised in the proposal: https://github.com/godotengine/godot-proposals/issues/11347

Changes made:

1) My new `AudioDriver::input_start_count` counter is to prevent multiple calls to `AudioDriver::input_start()` from different AudioStreamPlaybackMicrophone instances that result in multiple conflicting processes to popping data from the same buffer.

2) Change the return type of `AudioDriver::get_input_buffer()` to `Vector<int32_t>&` to avoid copying out entire buffer 
to access it.

3) `GDREGISTER_CLASS(AudioStreamPlaybackMicrophone)` was missing from `register_server_types.cpp`

4) Expose functions `start()`, `stop()` and `is_playing()` from `AudioStreamPlaybackMicrophone` to GDScript

5) New function `PackedVector2Array AudioStreamPlaybackMicrophone::get_microphone_buffer(int p_frames)`.  This does the same as [PackedVector2Array AudioEffectCapture::get_buffer(frames: int)](https://docs.godotengine.org/en/stable/classes/class_audioeffectcapture.html#class-audioeffectcapture-method-get-buffer) but it fetches the frames directly from the `AudioDriver::input_buffer` without going through an AudioStream, AudioBus and an AudioEffect, all of which operate on the duty cycle of the Audio system output frequency.

 
**Need help with:**

I would like to expose 
[int AudioStreamPlaybackMicrophone::mix(AudioFrame *p_buffer, float p_rate_scale, int p_frames)](https://github.com/godotengine/godot/blob/master/servers/audio/audio_stream.h#L254)
to GDExtension, but this has an `AudioFrame*` pointer in its parameters list  

This ought to be possible, because there is already one like this with [int AudioStreamPlaybackResampled::_mix_resampled(dst_buffer: AudioFrame*, frame_count: int)](https://docs.godotengine.org/en/stable/classes/class_audiostreamplaybackresampled.html#class-audiostreamplaybackresampled-private-method-mix-resampled), 
but it is created by the special virtual function template `GDVIRTUAL2R(int, _mix_resampled, GDExtensionPtr<AudioFrame>, int)`

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/11347*

### CODE REVIEW

**Comment by RedMser** [comment]
Since you modified the scripting API, make sure to update the class reference as well and fill it out: https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine

**Review by fire — CHANGES_REQUESTED** [review]
As we discussed, the design is that instead of the AudioServer using the mix callback somewhere else, I assume a node::_process(delta) will process the mic driver.

I am not convinced by the arguments to bypass the audio server to access the microphone driver directly.

**Comment by goatchurchprime** [comment]
**TLDR** 

There is a long and buggy route for getting microphone audio data out of the engine.  It would be shorter and more reliable to extract it directly from the AudioDriver's input_buffer.

**What happens now**

All microphone data goes into the ring-buffer [AudioDriver::input_buffer](https://github.com/godotengine/godot/blob/master/servers/audio_server.h#L62) advancing `input_position` as it is added.

The [AudioStreamPlaybackMicrophone](https://github.com/godotengine/godot/blob/master/servers/audio_server.h#L62) object has its own local pointer `input_ofs` into this same buffer, and its function [_mix_internal()](https://github.com/godotengine/godot/blob/master/servers/audio/audio_stream.cpp#L390) locks the audio device thread before returning a slice from this buffer using read access only.

At the moment the only way to extract microphone data from the engine is to wait for the AudioServer to request chunks of data from an AudioBus at its own update rate.  This AudioBus then requests the data from an AudioStreamPlaybackMicrophone filtered through an [AudioEffectCapture](https://docs.godotengine.org/en/stable/classes/class_audioeffectcapture.html#methods) object that copies any audio that passes through into its own little buffer.  Then, on a process loop, you extract the data from this buffer by calling [AudioEffectCapture.get_buffer()](https://docs.godotengine.org/en/stable/classes/class_audioeffectcapture.html#class-audioeffectcapture-method-get-buffer)

There are two problems with this.

Firstly, the AudioServer is highly sensitive to data not arriving in at the rate it requires, so if the microphone buffer isn't filled fast enough the stream is terminated.  There is a [pull-request](https://github.com/godotengine/godot/pull/93200) to fix the problem on the Android platform where the microphone keeps switching off after two minutes by padding the missing frames with zeros.  Not surprisingly, this degrades the quality of the microphone audio.

Secondly, we don't generally want any microphone audio data in the AudioServer because it causes feedback and is too laggy to work as realtime amplifier.  So if it is too difficult to debug this buffer properly (which it is), it's not worth it.  The evidence is that standard operating procedure requires us to create a special bus for the AudioStreamMicrophone to output to that is set to Mute.

**The solution**

Expose `AudioStreamPlaybackMicrophone::_mix_internal()` (suitably wrapped) as an external function, so that any process with a local copy of an `AudioStreamPlaybackMicrophone` can request slices of data from the `input_buffer` as it becomes available.  This makes no structural change to the engine, but allows a new access route to the microphone data avoiding the unnecessary requirement for it to operate in perfect synchronicity to the Audio streams.

**Comment by goatchurchprime** [comment]
This fix is working well.  It ran on my cheap Android phone for 4 hours without any issues at all.  

Previously it would last about 3 minutes on this phone before the `AudioStreamMicrophone` was switched off, often permanently.

Although there is some skepticism, I would like to push this PR into some place where it can be critically discussed and reviewed, because it actually works, and we have don't even have an outline for fixing the Microphone any other way. 

The primary function added into `AudioStreamPlaybackMicrophone` for accessing the audio data is:
```GDScript
PackedVector2Array get_microphone_buffer(int p_frames)
```

I would prefer to have the additional function:
```GDScript
bool mix_microphone(AudioFrame* buffer, int p_frames);
```

which would copy the values directly into an array rather than allocating and returning a temporary array by value.  

The main use of this interface is the [twovoip addon](https://github.com/goatchurchprime/two-voip-godot-4) for doing Voice over IP. 

There are a numerous virtual functions like [this one](https://docs.godotengine.org/en/stable/classes/class_audioeffectinstance.html#method-descriptions):
```GDScript
void _process(src_buffer: const void*, dst_buffer: AudioFrame*, frame_count: int) virtual
```

where the engine calls out to the GDExtension with an array pointer, but no examples I have found where a GDExtension calls into the engine with an array pointer.

My attempts to implement this function using:
```C++
bool AudioStreamPlaybackMicrophone::::mix_microphone(GDExtensionConstPtr<AudioFrame> p_buffer, int p_frames);
ClassDB::bind_method(D_METHOD("mix_microphone", "p_buffer", "frames"), &AudioStreamPlaybackMicrophone::mix_microphone);
```
do not compile, so I have left the call to the bind_method commented out until it's possible to find an answer.

**Comment by goatchurchprime** [comment]
The use of thread locking is inconsistent when accessing the [AudioDriver.input_buffer](https://github.com/godotengine/godot/blob/master/servers/audio_server.h#L62) ring-buffunctionfer.

In particular, values are pushed into this ring-buffer by the [AudioDriver::input_buffer_write()](https://github.com/godotengine/godot/blob/master/servers/audio_server.cpp#L104) function from various platforms without a mutex lock, while values are copied from this buffer by the function  [AudioStreamPlaybackMicrophone::_mix_internal()](https://github.com/godotengine/godot/blob/master/servers/audio/audio_stream.cpp#L389) where it is protected by a `lock()`.  

Therefore this locking achieves nothing, and we are getting away with it because either (1) the ring-buffer does not realloc, (2) there is a single audio thread on some platforms, or (3) any issues are mistaken for other bugs in the microphone system.

Confusingly, this ring-buffer has two indexes into it, `input_position` which is where the data is written, and `input_size` which is the number of elements in the buffer before the first wrap-around.  The conventional method of reading the microphone, using `AudioStreamPlaybackMicrophone::_mix_internal()`, ignores the `input_position` value and will happily overtake this index.  However, it uses the `input_size` to the buffer get ahead of the audio system by `playback_delay`= 50 miliseconds, which is often enough slack to allow things to run for a number of minutes before it [mysteriously stops](https://github.com/godotengine/godot/pull/93200).

My new [AudioStreamPlaybackMicrophone::get_microphone_buffer()](https://github.com/godotengine/godot/pull/100508/files#diff-e6f0ac593f0bd3b596ae48a5d3452a931b2d4acc7b553a1c4c346aad27c0c3c1R441) function however, does use the `input_position` value which, since it is not locked, could be encountered in an invalid state due to an interruption at [line 107](https://github.com/godotengine/godot/blob/master/servers/audio_server.cpp#L107) while `input_position >= input_buffer.size()` is true.

This lack of locking might explain the bizarre if-statement on [line 114](https://github.com/godotengine/godot/blob/master/servers/audio_server.cpp#L114) protecting this value from being invalid in this way that could only ever happen if two threads were executing this function simultaneously!  Such an event -- of two threads reading and posting microphone data into the buffer -- is already a bug, and would be fixed by Change 1 at the top. 

To solve this we either need to make this function safe for interruption, by rewriting it like (_I don't know if this is safe_ ? @RandomShaper ):
```C++
input_buffer.write[input_position] = sample;
input_position = (input_position % input_buffer.size());
```

or putting a lock into the `AudioDriver::input_buffer_write()` function,  which would be bad because it would mean the lock function was called up to 96,000 times per second.  

A more consistent answer would be to put the `lock()` into each of the 5 platform functions that copy values into this buffer, such as [AudioDriverOpenSL::_record_buffer_callback()](https://github.com/godotengine/godot/blob/bdf625bd54958c737fa6b7213b07581cc91059ad/platform/android/audio_driver_opensl.cpp#L186) for Android.

Since the exact time that we get data from `AudioStreamPlaybackMicrophone::get_microphone_buffer()` is not critical, it should be protected by a `trylock()` to avoid holding back the main worker thread.

**Comment by walksanatora** [comment]
person planning on adding VC to a game here.
from what I read (havent touched the PR) this would be a more stable and probally faster way of getting microphone input? which seems like it would be VERY usefull.
I could also imagine this being usefull for Speech To Text and reducing latency on that. (darn now I am considering Waltz Of Wizards verbal casting)

**Comment by goatchurchprime** [comment]
Further issues discovered:

1) The latest change to the Android `stop_input` doesn't leave it in a state where `start_input` can work again due to the pointers `recordItf` and `recordBufferQueueItf` not being set to null after their objects were deleted.
   https://github.com/godotengine/godot/pull/85955#issuecomment-2569938738
  I am now deleting the entire inputting system, including `recorder` to give the option of a completely fresh start.
 
 2) `AudioDriver::input_start()` takes an unreasonably long time on Windows, between 600 and 800ms, causing a freeze on startup.  Perhaps this should be moved into its own thread.
 
 3) The `try_lock()` in my `AudioStreamPlaybackMicrophone::get_microphone_buffer` returns false quite often on Android meaning it's doing some good by not blocking the main thread and instead deferring access to the microphone by a frame.

 3) The first run of an app on Android, where the system makes a popup so you can Grant Access to the Microphone leaves the microphone in a state where it cannot be restarted.  I'm debugging with a lot of print statements, but it looks like the `request_permission()` function is some kind of async, so that the code following this block gets executed after it's returned a fail to turn on the microphone, and then later turns it on when you accept the popup, leaving the higher level code in the believe it is still off.
 
```C++
 	if (!OS::get_singleton()->request_permission("RECORD_AUDIO")) {
		WARN_PRINT("Unable to start audio capture - No RECORD_AUDIO permission");
		return ERR_UNAUTHORIZED;
	}
```

**Comment by fire** [comment]
> The https://github.com/godotengine/godot/pull/85955#issuecomment-2569938738 to the Android stop_input() function leaves it in a state where start_input() cannot work again due to the pointers recordItf and recordBufferQueueItf not being set to null after their objects were deleted.

If you made a pr fixing this, you can get the android team to review it.

**Comment by goatchurchprime** [comment]
@m4gr3d @RandomShaper 

Need to review:

1. Consistent use of thread locking on the `AudioDriver.input_buffer` -- _or whether locking is actually necessary if we ensure there is only ever one writer on this ring-buffer that never resizes, and multiple independent readers that manage their own indexes._  (The [line 114 invalid state](https://github.com/godotengine/godot/blob/master/servers/audio_server.cpp#L114) can only happen if there are two non-locked threads writing to this buffer!)

2. Removing the automatic call to `AudioDriver.input_start()` following the user accepting Android Microphone Permissions that results in the higher level code not knowing when or if the microphone is working.  The logic of what to do after user acceptance or rejection can be kicked up to the GDScript level so that the game can know exactly why the microphone is not going to work.

3. The use of an instantiated copy of `AudioStreamPlaybackMicrophone` without its corresponding `AudioStreamMicrophone` to access chunks of audio data directly from the `AudioDriver.input_buffer` microphone buffer independently of the audio system.  (This does not change how anything else functions, because you can have multiple `AudioStreamPlaybackMicrophone`s referencing the single microphone buffer since each one has its own index.)

4. Proposed to change `AudioDriver.input_buffer` from an array of floats to an array of `AudioFrame`s.

**Comment by fire** [comment]
I am not happy with design of breaking the encapsulation of the audio system, it shouldn't be exposed like that.

**Comment by goatchurchprime** [comment]
This design does not break the encapsulation of the audio system!  The audio system is composed of busses, `AudioEffects` and `AudioStream` entities all connected to the output stream and being driven by its duty cycle.

When you instantiate an `AudioStreamPlaybackMicrophone` object in isolation it has no connection with the output stream.  If you prefer, we could duplicate the entire class with my functions and call it something else to make this fact more obvious, but its disconnection is enforced by the fact that its `Ref<AudioStreamMicrophone> microphone` member is null.

It is a common misconception that the audio input (microphone) is part of the audio system.  It clearly is not since it keeps running and doing its own thing even when the entire audio system has broken down or is working to a different frequency.  It is nothing more than a serial buffer receiving bytes from a serial device, often bluetooth serial.  

You can, however, attempt to connect its data directly into the audio system in real time, but (a) nobody wants this feature because it creates noisy feedback, and (b) the frequencies don't quite always match so any buffer will inevitably either over-run or starve.

I think people are very happy to treat the microphone input data as a data resource that can either be recorded for playback or compressed and transmitted over a network, especially if it means that it finally works and is stable.

**Comment by walksanatora** [comment]
finally: the ability to implement VoIP in godot games without doing very hacky work arounds.

**Comment by Ivorforce** [comment]
Superseded by https://github.com/godotengine/godot/pull/113288. Thanks again for all the hard work!
